File: src/middleware/guardian.ts
// src/middleware/guardian.ts
import { Request, Response, NextFunction } from 'express';
import { guardianStorage } from '../services/guardianStorage';
import { DetectionResult } from '../types';

export async function guardianMiddleware(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  const userAgent = req.headers['user-agent'] || 'Unknown';
  const ip = req.ip || req.socket.remoteAddress || 'Unknown';

  try {
    const result = await guardianStorage.detectBot({
      userAgent,
      ip,
      req
    });

    const detection: DetectionResult = {
      timestamp: new Date(),
      isBot: result.isBot,
      confidence: result.confidence,
      path: req.path || '/',
      userAgent,
      ip,
      reasons: result.reasons,
      behavior: result.behavior
    };

    await guardianStorage.storeDetection(detection);
    req.botDetection = detection;
  } catch (error) {
    console.error('Guardian middleware error:', error);
  }

  next();
}

File: src/client/python/guardian.py
import requests
import json
from functools import wraps

class GuardianClient:
    def __init__(self, api_url='http://localhost:3333'):
        self.api_url = api_url
        
    def detect_bot(self, user_agent, ip, path=None):
        """Detect if a request is from a bot"""
        response = requests.post(
            f"{self.api_url}/detect",
            json={
                "userAgent": user_agent,
                "ip": ip,
                "path": path
            }
        )
        return response.json()
    
    # Flask middleware
    def middleware(self):
        """Flask middleware for bot detection"""
        def decorator(f):
            @wraps(f)
            def decorated_function(*args, **kwargs):
                from flask import request, g
                
                result = self.detect_bot(
                    user_agent=request.headers.get('User-Agent', ''),
                    ip=request.remote_addr,
                    path=request.path
                )
                g.bot_detection = result
                return f(*args, **kwargs)
            return decorated_function
        return decorator